### PR TITLE
feat: package cache

### PR DESCRIPTION
**Pull request by baszalmstra**
Provides functionality to cache extracted Conda packages in a directory. The cache object holds an in-memory cache of which valid packages exist. It checks the filesystem if an entry doesn't exist in memory yet. Validation (from #39 ) is used to determine whether the cache is corrupt. If the filesystem also doesn't provide the requested package it is fetched into the cache.

The cache does not provide a method to download and extract package archives but instead relies on a `populate` function. The goal is to decouple "caching" and "download/extract". This makes it easier to test, provide different downloading functionalities, queueing, etc.

Builds on #39 which should be merged first!

### CODE REVIEW

**Comment by baszalmstra** [comment]
@wolfv Do you have ideas for edge cases or different test strategies?

File: crates/rattler/src/validation.rs [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+    };
+
+    // Validate all the entries
+    validate_package_directory_from_paths(package_dir, &paths)
+        .map_err(|(path, err)| PackageValidationError::CorruptedEntry(path, err))?;
+
+    Ok(paths)
+}
+
+/// Determine whether the files in the specified directory match wat is expected according to the
```
Comment (wolfv) [inline crates/rattler/src/validation.rs:96]: small typo (wat -> what).

**Review by wolfv — APPROVED** [review]

**Comment by wolfv** [comment]
I think the biggest edge cases I have in mind are all around:

- file locking when writing to the cache from multiple processes (especially when network file systems are involved)
- layered caches (e.g. one in `/usr/cache` that is read-only, and one in `~/cache` that is read-write)
- finding a writable cache

**Review by wolfv — APPROVED** [review]
